src/range.cr

```diff
@@ -497,6 +497,12 @@ struct Range(B, E)
     end
   end
 
+  # Returns `self`.
+  # A `Range` does not contain any `Nil` elements by definition.
+  def compact : self
```

Comment:
This should return an Array.

Reply:
Or put another way: this method for Range makes no sense.

Reply:
Yes, that's true. But this is actually not specific to `Range`. `Enumerable(T)#compact` where `T` is not nilable is always a no-op and returns `self` because there's nothing to remove. That's implemented in `Enumerable#compact` with a macro condition on `T >= Nil`.
The override in `Range#compact` is necessary because a left-open range (`Range(Nil, E) forall E`) is *technically* typed as `Enumerable(Nil)`. But semantically it doesn't contain any `Nil` element, so `#compact` is still a no-op.

Reply:
This is what I mean:

```
$ irb
irb(main):001:0> (1..10).compact
=> [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

I don't understand why `compact` for an `Enumerable` would return something that's not an Array.

I understand for Set, Dequeue and other **collection** types it might make sense to return that collection type. But Range is not a collection.

In any case, feel free to ignore this comment. I think it's all related to my regret of having `Tuple`, `StaticArray`, `Deuque`, etc.

Heck, even in Ruby calling `compact` on `Set` returns an `Array`.

Reply:
A beginless `Range` doesn't have "no" elements, because any attempt to iterate on it raises immediately before even returning. So if `(..3).compact` would have raised if it were to return an `Array`, it too should raise if it returns a `Range`. This current implementation doesn't check that yet.

What I also expect here is that calling `Enumerable(T)#compact` should remove `Nil` from the `T`, regardless of the concrete type, so `(1.as(Int32?)..4).compact` should return either `1..4` or `[1, 2, 3, 4] of Int32`. Again this overload doesn't do that.

Reply:
Right, I think it's a bit strange for `compact` to exist on Range. I think `compact` didn't exist in Enumerable in previous Ruby versions either.

It seems `(1..).compact` in Ruby just hangs.

Reply:
> So if `(..3).compact` would have raised if it were to return an `Array`, it too should raise if it returns a `Range`.

I disagree. The problem lies with the definition that `..3` is an `Enumerable(Nil)`. That's just not true. This range does not contain any `Nil` elements. It's actually a collection of integers, and the correct inheritance type would be `Enumerable(Int32)`. But it's impossible to express that in the type grammar 😢 
There's no way to iterate it from the start (because it has no start), but it works from the end: `(..3).reverse_each` yields 3, 2, 1, 0, -1, ... etc.
And `(..3).compact.reverse_each` should do *exactly* the same. `compact` is supposed to remove `Nil` elements. If there are none: good job, nothing to do.

Reply:
It would be `Enumerable(NoReturn)` at best, not `Enumerable(Int32)`. What `#reverse_each` does is irrelevant to the semantics of `Enumerable` alone, and we don't have a module for `#reverse_each` yet.

Reply:
I guess it's fine then to return `self`. I'm just thinking about users calling `compact` on something and not really knowing what to expect. Will it be an Array? A Set? A Range? Who knows! But I guess nobody would really call `compact` on a Range, so it's fine.

Reply:
`#reverse_each` is just meant as a concrete demonstration for the item type of an open-ended range. I mean it clearly *has* items. For example: `(..3).includes?(2) # => true`. And they have a type. `Int32` in this case.
Neither `Enumerable(Nil)` nor `Enumerable(NoReturn)` make sense as describing the item type.
I suppose `Enumerable(NoReturn)` could be considered as an indication that the collection is not iterable from the start. It would definitely be better than `Enumerable(Nil)`. Still, we can't really express that.

Reply:
@asterite 
> Will it be an Array? A Set? A Range? Who knows! 

The API docs know =)

Reply:
`(1..6).includes?(2.345)` is also true; `Range#includes?` treats the receiver like a distribution rather than a collection, and does _not_ respect `Enumerable#includes?`'s semantics. So `#includes?` here isn't really a good example.

Reply:
Let's go a step back:

> `Enumerable(T)#compact` where `T` is not nilable is always a no-op and returns `self` because there's nothing to remove.

If we follow this then every `Hash` would simply return itself, because it is an `Enumerable(Tuple(K, V))`, and `Tuple` is obviously non-nilable.

So maybe it really doesn't make sense to generalize `#compact` to all `Enumerable`s...?

Reply:
> `(1..6).includes?(2.345)` is also true

I'd say this is more of a bug of `Range`.